overlore/sqlite/vector_db.py

```diff
@@ -0,0 +1,135 @@
+from sqlite3 import Connection
+from typing import Any
+
+import sqlite_vss
+
+from overlore.prompts.prompts import GptInterface
+from overlore.sqlite.db import Database
+from overlore.sqlite.types import StoredVector, TownhallEventResponse
+
+
+def save_to_file(discussion: str, rowid: int, embedding: str) -> None:
+    with open("Output.json", "a") as file:
+        data_to_save = {"rowid": rowid, "discussion": discussion, "embedding": embedding}
+        json.dump(data_to_save, file)
+        file.write("\n")
+
+
+class VectorDatabase(Database):
+    _instance: VectorDatabase | None = None
+    GPT_CONN: GptInterface | None = None
+    EXTENSIONS: list[str] = []
+    # EXTENSIONS: list[str] = ["vector0", "vss0"]
+    FIRST_BOOT_QUERIES: list[str] = [
+        """
+            CREATE TABLE IF NOT EXISTS townhall (
+                discussion text,
+                realmID int,
+                events_ids text,
+                ts text
+            );
+        """,
+        """
+            CREATE VIRTUAL TABLE vss_townhall using vss0(
+                embedding(1536)
+            );
+        """,
+    ]
+
+    @classmethod
+    def instance(cls) -> VectorDatabase:
+        if cls._instance is None:
+            print("Creating vector db interface")
+            cls._instance = cls.__new__(cls)
+        return cls._instance
+
+    def __init__(self) -> None:
+        raise RuntimeError("Call instance() instead")
+
+    def _preload(self, db: Connection) -> None:
+        sqlite_vss.load(db)
+
+    def init(self, path: str = "./vector.db", gpt_interface: GptInterface | None = None) -> VectorDatabase:
+        # Call parent init function
+        self._init(path, self.EXTENSIONS, self.FIRST_BOOT_QUERIES, [], self._preload)
+        self.GPT_CONN = gpt_interface
+        return self
+
+    def vss_version(self) -> str:
+        cursor = self.db.cursor()
+        (version,) = cursor.execute("select vss_version()").fetchone()
+        return str(version)
+
+    def mock_insert(self, data: dict[str, Any]) -> None:
+        ts = data["ts"]
+        rowid = self._insert(
+            "INSERT INTO townhall (discussion, realmID, events_ids, ts) VALUES (?, ?, ?, ?);",
+            (data["discussion"], data["realmID"], json.dumps(data["events_ids"]), ts),
+        )
+
+        embedding = data["embedding"]
+        self._insert("INSERT INTO vss_townhall(rowid, embedding) VALUES (?, ?)", (rowid, json.dumps(embedding)))
+
+    def get_all(self) -> tuple[int, int]:
+        query = "SELECT * FROM townhall"
+        records = self.execute_query(query, ())
+
+        vss_query = "SELECT * FROM vss_townhall"
+        records_vss = self.execute_query(vss_query, ())
+
+        return len(records), len(records_vss)
+
+    async def insert_townhall_discussion(
+        self, discussion: str, realmId: int, event_ids: str, save: bool = False
+    ) -> None:
+        discussion = discussion.strip()
+        ts = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
```

Comment:
im thinking it MIGHT be good to couple generated townhall w/ current block of katana. doesn't necessarily need to happen in this PR

Reply:
Sounds like an idea, I don't really have a use for it rn but let's not forget it

Reply:
Created an issue for this.